hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/HoodieTimelineArchiver.java

```diff
@@ -515,13 +515,25 @@ private Stream<HoodieInstant> getInstantsToArchive() {
   private boolean deleteArchivedInstants(List<HoodieInstant> archivedInstants, HoodieEngineContext context) throws IOException {
     LOG.info("Deleting instants " + archivedInstants);
     boolean success = true;
-    List<String> instantFiles = archivedInstants.stream().map(archivedInstant ->
-        new Path(metaClient.getMetaPath(), archivedInstant.getFileName())
```

Comment:
The `archivedInstants` seems already been sorted, because the instants from the timeline are already sorted by the instant timestamp and **state**, while in 
https://github.com/apache/hudi/blob/b10ca7e69f43cf2f14b09d0a610102ab058b1511/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/HoodieTimelineArchiver.java#L465

we group by the instant pairs but it should still keep sequence for single instant,
one instant would be sorted by state: requested -> inflight -> complete,
so in theory, the files are also cleaned in this sequence.

Reply:
Yes, that's right. But I'm not sure if `deleteFilesParallelize` is cleaned in order. It use `FSUtils.parallelizeFilesProcess` to delete file and depend on implement in different engine.

https://github.com/apache/hudi/blob/b10ca7e69f43cf2f14b09d0a610102ab058b1511/hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/client/common/HoodieFlinkEngineContext.java#L145-L147

For example, in flink engine context, it use parallel stream to delete file. 
This is my understanding, please ping me if I am wrong.

Reply:
Yes, the parallel deletion may cause the disorder of the files.

But there is still a confusion needs to be resolved here:
The archiver only archives the old/completed instant on the timeline, while the flink writer only checks the latest inflight instant which should not be archived yet, so why the flink client is affected here ?

Reply:
I had this happen last time. 
Flink Application's JobManager was force killed because of memory. 
Ant it left some instant file:

* 20220513170152006.commit.requested
* 20220513170152006.inflight
* 20220513170202566.commit
* 20220513170202566.commit.requested
* 20220513170202566.inflight

In this case, flink Application will execute this code:

https://github.com/apache/hudi/blob/7fb436d3cf66748f32776cf3db2d943f97b52160/hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/meta/CkpMetadata.java#L97-L102

It will scan and get last pending instant, and create a in-flight file in `.hoodie/.aux/ckp_meta`(`20220513170152006` in this case).

The final result will throw an excpetion: `Found commits after time :20220513170152006, please rollback greater commits first`

https://github.com/apache/hudi/blob/7fb436d3cf66748f32776cf3db2d943f97b52160/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/rollback/BaseRollbackActionExecutor.java#L179-L183

Reply:
Cool, the logic to fetch the pending instant is problematic, but the error trace you paste is mainly because when collecting the instants to rollback in line:

https://github.com/apache/hudi/blob/7fb436d3cf66748f32776cf3db2d943f97b52160/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieWriteClient.java#L1084

We do not consider that there are inflight instants (previously completed but inflight again because of the archiving) on the timeline, cc @nsivabalan :)

Reply:
I have a patch to fix the pending instant time fetch but it does not solve your problem :)
[fix.patch.zip](https://github.com/apache/hudi/files/8686763/fix.patch.zip)

Reply:
@nsivabalan :)